### ISSUE TITLE
Enhancement: themes for menu items

### ISSUE THREAD

**Issue by AlexeiBugai** [vaadin/vaadin-menu-bar-flow#34]
Hi!
1. Can you add ThemeVariants like LUMO_SUCCESS and LUMO_ERROR also?
2. Is it possible to set ThemeVariants to MenuItem? Not to whole MenuBar like now.
Thank you.

**Comment by pekam** [comment]
Hi!

1. We left these theme variants out intentionally, as we thought that they are not useful for menu-bar. So we didn't want to add extra stuff to the API and encourage users to set `success` or `error` theming for the menu-bar. All the theme variants of `vaadin-button` work however, so as a workaround you can set it with the Element API:
`menuBar.getElement().setAttribute("theme", "success");`
2. We have not considered this kind of an use case either. We think that usually the menu-bar buttons should be consistently styled. Again, I can provide a workaround by using the Element API:
`menuItem.getElement().executeJs("this.parentElement.setAttribute('theme', 'error')");`

Can you tell us about your use case, so we can maybe re-evaluate if the `error` and `success` variants and themes for individual buttons would be useful for menu-bar?

**Comment by AlexeiBugai** [comment]
I have implemented the next logic:
- buttons which change something in DB (like save form) have own `ThemeVariant`.
- buttons which have some shortkey (default buttons for form or grid) also have own `ThemeVariant`.
- for some important buttons I can use `error` with red color.
So, form can have big menu with a lot of buttons, and most important buttons have special color...

**Comment by pekam** [comment]
If I understood correctly, it sounds like you are using menu-bar as a row of buttons. For this kind of case you should use the `Button` component.

The main use case for menu-bar is to create similar menus with sub-menus as you can see on the top of most desktop applications.

But if you need to control each button of `MenuBar`, you can create separate `MenuBars` for each item (and use `Button` when you don't need a sub-menu), although this is clearly not optimal.

**Comment by heruan** [comment]
I have another use case: composite buttons. For example, I'd like to achieve this:

<img width="319" alt="Screenshot 2019-05-24 at 15 53 33" src="https://user-images.githubusercontent.com/851745/58332736-270fdf80-7e3c-11e9-900a-4fa25638bdd6.png">

but this code

```java
var bar = new MenuBar();
bar.addThemeVariants(MenuBarVariant.LUMO_PRIMARY);
bar.addItem("Send as Working");
var chevron = bar.addItem(VaadinIcon.CHEVRON_DOWN.create());
var submenu = chevron.getSubMenu();
submenu.addItem("Send as Pending");
submenu.addItem("Send as Closed");
```

produces this instead

<img width="314" alt="Screenshot 2019-05-24 at 15 53 47" src="https://user-images.githubusercontent.com/851745/58332744-2d9e5700-7e3c-11e9-939d-de428d511754.png">

and `chevron.getElement().getThemeList().add("icon")` appearantly does nothing.

**Comment by heruan** [comment]
See also https://github.com/vaadin/vaadin-button/issues/86

**Comment by heruan** [comment]
I've also tried

```java
chevron.getElement().executeJs("this.parentElement.setAttribute('theme', 'icon')");
```

as suggested by @pekam but it didn't work either.

**Comment by pekam** [comment]
Looks like there's some timing issue which can be avoided with `setTimeout`:
```java
chevron.getElement().executeJs(
  "setTimeout(function(){this.parentElement.setAttribute('theme', 'icon');}.bind(this))");
```
But this workaround is getting pretty ugly.

**Comment by Haprog** [comment]
I don't think it's a good idea to implement that kind of composite buttons as separate items in a `vaadin-menu-bar`. You're much better off using a some separate component for the composite (which could use e.g. `vaadin-context-menu` for the submenu part) and use that as a single item if you need it in the menu bar.

For example if you have many items or narrow viewport, with your above approach of having two separate items, the chevron button can become detached from the other button and go under the overflow menu item ("...") and then it would seem quite broken imo.

**Comment by pekam** [comment]
> You're much better off using a some separate component for the composite (which could use e.g. vaadin-context-menu for the submenu part)

MenuBar is currently the only component that supports this kind of overlay positioning for a button out of the box, and thus it's the best solution for this use case at the current state of our components (imo).

**Comment by heruan** [comment]
@Haprog I agree, and I've already tried to build a custom button with `vaadin-context-menu` but the alignment was driving me crazy.

So I guess my use case needs to wait for some other component as @jouni suggested in https://github.com/vaadin/vaadin-button/issues/86#issuecomment-465263155 and mockups in https://github.com/vaadin/vaadin-button/issues/86#issuecomment-387481409

**Comment by Haprog** [comment]
> MenuBar is currently the only component that supports this kind of overlay positioning for a button out of the box, and thus it's the best solution for this use case at the current state of our components (imo).

Seems like this kind of overlay positioning support would be needed for other components too then. Maybe at least `vaadin-context-menu`. Related: vaadin/vaadin-context-menu#153.

@heruan Yes I think that feature vaadin/vaadin-button#86 would best suit this use case. But a smaller step to help with the implementation would probably be if vaadin-context-menu had some option to configure the alignment to be stuck to the side of a parent/target element.

@heruan Have you tried if this workaround could work for you to position the context menu e.g. under a button to build your own composite button https://github.com/vaadin/vaadin-context-menu/issues/153#issuecomment-455483333 ? If that's not enought then at this point I guess the only choice would be to implement your own component and maybe extend or reuse some of our components where it helps.

**Comment by pekam** [comment]
Valid points, but let's not use this ticket to design a new composite button component, for which we have https://github.com/vaadin/vaadin-button/issues/86.

Setting e.g. `theme='icon'` for a menu-bar button still seems like a generally useful feature for menu-bar when one of the buttons contains just an icon. Still not sure about other variants.

A menu-bar could still be a valid solution for a similar use case as a "composite button", although these two components have clear differences.

**Comment by heruan** [comment]
Right, I'll follow-up on the other issue for the composite button! This is still useful for applying themes to menu items, thank you @pekam!

**Comment by AlexeiBugai** [comment]
Hi Pekka!
I used the code as you suggested, but it does not work.
menuItem.getElement().executeJs("this.parentElement.setAttribute('theme', 'error')");
Something wrong with executeJs in beta2?

**Comment by Haprog** [comment]
@AlexeiBugai I guess you didn't see this https://github.com/vaadin/vaadin-menu-bar-flow/issues/34#issuecomment-496397679

**Comment by AlexeiBugai** [comment]
Thanks @Haprog , but this option also don't work for me. Am I doing something wrong?
`menuItem.getElement().executeJs("setTimeout(function(){this.parentElement.setAttribute('theme', 'primary');}.bind(this))");`

**Comment by Enerccio** [comment]
Maybe it is late, but what I did to style item is to copy paste ALL classes from menu (ContextMenu, MenuBar, MenuItem, HasMenuItems, SubMenu), renamed them and then refactor them to apply to each other. Then I could include JsModule on my copy of MenuItem. And to remove justify I did:

```$_documentContainer.innerHTML = `
<dom-module id="my-menuitem-button" theme-for="vaadin-context-menu-item">
  <template>
    <style>
        [part="content"] {
            justify-content: left !important;
        }
    </style>
  </template>
</dom-module>
`;```

This is incredibly convoluted and should be easier, ffs!

**Comment by jouni** [comment]
> I don't think it's a good idea to implement that kind of composite buttons as separate items in a vaadin-menu-bar. You're much better off using a some separate component for the composite (which could use e.g. vaadin-context-menu for the submenu part) and use that as a single item if you need it in the menu bar.

> For example if you have many items or narrow viewport, with your above approach of having two separate items, the chevron button can become detached from the other button and go under the overflow menu item ("...") and then it would seem quite broken imo.

Late to comment, but I have to disagree. I think Menu Bar could be a suitable workaround for split buttons. In such cases, it should be possible to disabled the default collapsing/overflow behavior (possible at least by specifying a min-width that can fit both buttons).

**Comment by Haprog** [comment]
> I think Menu Bar could be a suitable workaround for split buttons.

I agree it can be a functional _workaround_, but imo it's not ideal and would be good to have a better solution that also works with the overflow collapsing.

**Comment by jouni** [comment]
Sure. Basically having a way to indicate that certain items should always be collapsed into the overflow menu.
